src/dune/opam_create.ml

```diff
@@ -130,6 +131,33 @@ let insert_dune_dep depends dune_version =
   in
   loop [] depends
 
+let rec already_requires_odoc : Package.Dependency.Constraint.t -> bool = function
+  | Bvar (Var "with-doc") | Uop (_, QVar _) -> true
+  | Bvar _ | Uop _ | Bop _ -> false
+  | And l -> List.for_all ~f:already_requires_odoc l
+  | Or l -> List.exists ~f:already_requires_odoc l
```

Comment:
To insert the dependency, we have a simpler heuristic of just not touching the constraint if the user wrote something by hand. I wonder if we should keep it simple here as well?